xayn-ai/src/embedding/utils.rs

```diff
@@ -57,23 +57,32 @@ where
 ///
 /// # Panics
 /// Panics if the vectors don't consist solely of real values or their shapes don't match.
+///
+/// # Safety
+/// Only use this with [`TrustedLen`] iterators. The bound can't be named currently, because the
+/// trait is nightly-gated. [`ExactSizeIterator`] isn't a feasible replacement, because it isn't
+/// implemented for [`Chain`]ed iterators for example.
+///
+/// [`TrustedLen`]: std::iter::TrustedLen
+/// [`ExactSizeIterator`]: std::iter::ExactSizeIterator
+/// [`Chain`]: std::iter::Chain
 #[allow(dead_code)]
-pub fn pairwise_cosine_similarity<'a, I, S>(iter: I) -> Array2<f32>
+pub unsafe fn pairwise_cosine_similarity<I, S>(iter: I) -> Array2<f32>
 where
-    I: IntoIterator<Item = &'a ArrayBase<S, Ix1>>,
-    I::IntoIter: Clone + ExactSizeIterator,
-    S: Data<Elem = f32> + 'a,
+    I: IntoIterator<Item = ArrayBase<S, Ix1>>,
+    I::IntoIter: Clone,
+    S: Data<Elem = f32>,
 {
     let iter = iter.into_iter();
-    let size = iter.len();
+    let size = iter.size_hint().0;
 
     let norms = iter.clone().map(|a| l2_norm(a.view())).collect::<Vec<_>>();
     let mut similarities = Array2::ones((size, size));
     for (i, a) in iter.clone().enumerate() {
         if norms[i] != 0. {
             for (j, b) in iter.clone().enumerate().skip(i + 1) {
                 if norms[j] != 0. {
-                    similarities[[i, j]] = a.dot(b) / norms[i] / norms[j];
+                    similarities[[i, j]] = a.dot(&b) / norms[i] / norms[j];
```

Comment:
As this uses bounds checks we don't really need to  make the function `unsafe` a bad `size` will just panic.

Reply:
removed the `unsafe`, updated the docs and calls